src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,360 @@
+(ns com.fulcrologic.fulcro.algorithms.normalized-state-helpers
+  "Functions that can be used against a normalized Fulcro state database."
+  #?(:cljs (:require-macros com.fulcrologic.fulcro.algorithms.normalized-state-helpers))
+  (:refer-clojure :exclude [get-in])
+  (:require
+    [clojure.walk :as walk]
+    [com.fulcrologic.fulcro.components :as comp :refer [defsc]]
+    #?(:clj  [com.fulcrologic.fulcro.dom-server :as dom]
+       :cljs [com.fulcrologic.fulcro.dom :as dom])
+    [edn-query-language.core :as eql]
+    [clojure.spec.alpha :as s]
+    [ghostwheel.core :refer [>defn =>]]
+    [com.fulcrologic.fulcro.algorithms.data-targeting :as targeting]
+    [com.fulcrologic.fulcro.algorithms.denormalize :as fdn]
+    [com.fulcrologic.fulcro.algorithms.merge :as merge]
+    [com.fulcrologic.fulcro.components :as comp]))
+
+
+
+;;;;;;;;;;;;;;;;;;
+;; TODO
+
+;; - [ ] Test this namespace works in CLJC
+;; - [ ]  Learn how to run the ghostwheel test in the browser
+;; - [ ] Try out generative testing
+
+
+;;============================================================================
+
+(>defn tree-path->db-path
+  "Convert a 'denormalized' path into a normalized one by walking the path in state and honoring ident-based edges.
+
+  For example, one might find this to be true for a normalized db:
+
+  ```
+  state => {:person/id {1 {:person/id 1 :person/spouse [:person/id 3]}
+                        3 {:person/id 3 :person/first-name ...}}}
+
+  (tree-path->db-path state [:person/id 1 :person/spouse :person/first-name])
+  => [:person/id 3 :person/first-name]
+  ```
+  "
+  ([state path]
+   [map? vector? => vector?]
+   (tree-path->db-path state path path))
+
+  ([state path not-found]
+   [map? vector? any? => vector?]
+   (loop [[h & t] path
+          new-path []]
+     (if h
+       (let [np (conj new-path h)
+             c (clojure.core/get-in state np)]
+         (if (eql/ident? c)
+           (recur t c)
+           (recur t (conj new-path h))))
+       (if (not= path new-path)
+         new-path
+         not-found)))))
+
+
+
+(comment
+
+  '())
+
+
+;;============================================================================
+
+(>defn get-in
+  "Just like clojure.core/get-in, but if an element of the path is an ident it will follow the ident instead."
+  ([state-map path]
+   [map? vector? => any?]
+   (get-in state-map path nil))
+
+  ([state-map path not-found]
+   [map? vector? any? => any?]
+   (clojure.core/get-in state-map (tree-path->db-path state-map path) not-found)))
+
+(comment
+
+  ;;;;;
+
+  '())
+
+
+
+;;============================================================================
+
+
+
+(defn- paths [m]
```

Comment:
I've zeroed in on this one to be a good approach which covers all cases. 

- Get all possible paths in the app-db 
- Remove the idents ( if present) from the toplevel ( map and to-one key) 
- Filter out the paths which are more than 4 in length ( means they are denormalized)
- Run the check for idents and update on the remaining paths 
     This requires multiples updates to the state map to be done, which doesn't seem like the right way and I'm trying to make this more idiomatic.

Reply:
Walkthrough of the approach

- Example state atom 

```clojure
             {:a      [:b 1]
              :top-tm [[:d 1] [:b 1]]
              :denorm {:level-1 {:level-2 {:a [[:b 1]] :b [:b 1]}}}
              :b      {1 {:c [:d 1]
                          :e [[:b 1] [:d 1]]}}
              :d      {1 {:value 42}}}
```

- All paths and correspoding value (via `nsh/get-in`)

```clojure 
{[:d 1 :value]                  42,
 [:b 1 :e]                      [[:b 1] [:d 1]],
 [:b 1 :c]                      {:value 42},
 [:denorm :level-1 :level-2 :b] {:c [:d 1], :e [[:b 1] [:d 1]]},
 [:denorm :level-1 :level-2 :a] [[:b 1]],
 [:top-tm]                      [[:d 1] [:b 1]],
 [:a]                           {:c [:d 1], :e [[:b 1] [:d 1]]}}

```

- Remove the toplevel `ident` if it exists - say `[:b 1]`

```clojure
{[:d 1 :value]                  42,
 [:denorm :level-1 :level-2 :b] {:c [:d 1], :e [[:b 1] [:d 1]]},
 [:denorm :level-1 :level-2 :a] [[:b 1]],
 [:top-tm]                      [[:d 1] [:b 1]],
 [:a]                           {:c [:d 1], :e [[:b 1] [:d 1]]}}
```

- Reconstruct the modified `app-db` via `reduce`
    + Leave out the denormalized data, by a check on the length of path.
    + Remove the dangling pointers to that value in all `to-one` and `to-many` collections